java/code/src/com/redhat/rhn/manager/audit/CVEAuditManagerOVAL.java

```diff
@@ -0,0 +1,295 @@
+            // implementation.
+            // TODO: We could make a custom query that would get us only the data we're interested in instead of relying
+            //  on CVEAuditManager#doAuditSystem implementation.
+
+            CVEAuditSystemBuilder auditWithChannelsResult =
+                    CVEAuditManager.doAuditSystem(clientServer.getId(), resultsBySystem.get(clientServer.getId()));
+
+            systemAuditResult = doAuditSystem(cveIdentifier, resultsBySystem.get(clientServer.getId()),
+                    clientServer);
+            systemAuditResult.setChannels(auditWithChannelsResult.getChannels());
+            systemAuditResult.setErratas(auditWithChannelsResult.getErratas());
+
+            if (patchStatuses.contains(systemAuditResult.getPatchStatus())) {
+                result.add(new CVEAuditServer(
+                        systemAuditResult.getId(),
+                        systemAuditResult.getSystemName(),
+                        systemAuditResult.getPatchStatus(),
+                        systemAuditResult.getChannels(),
+                        systemAuditResult.getErratas()));
+            }
+        }
+
+        return result;
+    }
+
+    private static boolean isCVEIdentifierUnknown(String cveIdentifier) {
+        return !OVALCachingFactory.canAuditCVE(cveIdentifier);
+    }
+
+    /**
+     * Audit the given {@code clientServer} regarding the given CVE identifier based on OVAL and Channels data.
+     *
+     * @param clientServer the server to audit
+     * @param results list produced by {@link CVEAuditManager#listSystemsByPatchStatus(User, String)},
+     *                helpful for determining the availability of a patch for the vulnerability in channels.
+     * @param cveIdentifier the CVE identifier
+     * @return a record with data about a single system containing that system's patch status regarding a certain
+     * given CVE identifier as well as sets of relevant channels and erratas.
+     * */
+    public static CVEAuditSystemBuilder doAuditSystem(String cveIdentifier,
+                                                      List<CVEAuditManager.CVEPatchStatus> results,
+                                                      Server clientServer) {
+        // It's possible to have 2 or more patches for one package. It's necessary to apply all of them because
+        // they will have the same outcome i.e. patch the package; instead we need to choose only one.
+        // To choose the one, we rank patches based on the channel they come from .e.g.
+        // assigned, successor product, etc. And for each vulnerable package we keep only the highest ranking patch
+        results = keepOnlyPatchCandidates(results);
+
+        CVEAuditSystemBuilder cveAuditServerBuilder = new CVEAuditSystemBuilder(clientServer.getId());
+        cveAuditServerBuilder.setSystemName(clientServer.getName());
+
+        List<ShallowSystemPackage> allInstalledPackages =
```

Comment:
If we only care about installed packages that are affected by the CVE should we load only the packages affected but the CVE that are installed, and this way avoid loading all packages to memory?

Reply:
I'm all for the idea. There is already a query in PR D that could help in that regard. I think we can delay this until PR D is reviewed and take advantage of that query, this is especially useful to not create confusing conflicts with changes in the other PRs.


Query in PR D:
https://github.com/uyuni-project/uyuni/blob/4f5a7f459e7ff0e9577698c3f5a76e910d656790/java/code/src/com/redhat/rhn/common/db/datasource/xml/oval_queries.xml#L33